contents/blog/envoy-wizard-llm-agent.md

```diff
@@ -0,0 +1,75 @@
+---
+title: "Join the LLM coding agent revolution for $0"
+date: 2025-04-23
+author:
+  - danilo-campos
+featuredImage: >-
+  https://res.cloudinary.com/dmukukwp6/image/upload/posthog.com/contents/images/blog/posthog-engineering-blog.png
+featuredImageType: full
+tags:
+    - Engineering
+---
+
+AI agents are taking over the workflows of software development. They can burn through boilerplate and tedium fast, getting working prototypes up in minutes.
+
+The first problem with LLM codegen: there are so many stupid ways to build software.
+
+When an agent sets out to write code, it has many viable paths to both solve the problem and introduce mistakes. Mistakes are especially common for newer, upstart projects.
+
+Scale is one way to compensate: fill the training set with enough statistical weight that favors correct implementations, you’re less likely to see the boneheaded stuff.
+
+But that’s a lot of effort, plus significant lag time, and you want correct agent behavior *today*.
+
+The second problem with LLM codegen: it's very expensive. So what if you could join this revolution in software development without constructing an elaborate donation machine to Anthropic's coffers?
+
+You can! Build an **envoy**: software you control that developers can inject into their agent sessions with just a prompt.
+
+See, **these agents can run terminal commands**, so any CLI-based program can adopt this pattern.
+
+With this deterministic surface introduced to the agent workflow, you can reliably and predictably:
+
+- Enforce progression conditions, not moving to another step until requirements are satisfied  
+- Enforce correct versioning
+- Use API interactions to populate account secrets and other details, while enforcing good version control hygiene  
+- Introduce specific fragments of code verbatim
+- Inject up-to-date documentation and agent rules into the project to course-correct future agent sessions  
+- Do your own LLM-based transformations on code, using prompts, context and model versions that are known to be reliable for your goals
+- Enjoy same-day updates to agent workflows, just like the rest of your software
+
+Think of it like a sheepdog that keeps the agent session within a territory that is productive rather than stupid, adhering to best practices **you** as a developer can describe in code.
```

Comment:
```suggestion
Think of it like a sheepdog that keeps the agent session within a territory that is productive rather than stupid, adhering to best practices **you** as a developer can describe in code.

Naturally, we're building our own envoy already. It's still an alpha, but it's already very cool and we have big plans for it.
```